internal/dataplane/parser/translators/httproute_atc.go

```diff
@@ -234,55 +233,47 @@ func matchersFromParentHTTPRoute(hostnames []string, metaAnnotations map[string]
 	return ret
 }
 
-const (
-	InternalRuleIndexAnnotationKey  = "internal-rule-index"
-	InternalMatchIndexAnnotationKey = "internal-match-index"
-)
+type SplitHTTPRouteMatch struct {
+	Source     *gatewayv1beta1.HTTPRoute
+	Hostname   string
+	Match      gatewayv1beta1.HTTPRouteMatch
+	RuleIndex  int
+	MatchIndex int
+}
 
-// SplitHTTPRoute splits HTTPRoutes into HTTPRoutes with at most one hostname, and at most one rule
+// SplitHTTPRoute splits HTTPRoutes into matches with at most one hostname, and one rule
 // with exactly one match. It will split one rule with multiple hostnames and multiple matches
 // to one hostname and one match per each HTTPRoute.
-func SplitHTTPRoute(httproute *gatewayv1beta1.HTTPRoute) []*gatewayv1beta1.HTTPRoute {
-	// split HTTPRoutes by hostname.
-	hostnamedRoutes := []*gatewayv1beta1.HTTPRoute{}
-	if len(httproute.Spec.Hostnames) == 0 {
-		hostnamedRoutes = append(hostnamedRoutes, httproute.DeepCopy())
-	} else {
-		for _, hostname := range httproute.Spec.Hostnames {
-			hostNamedRoute := httproute.DeepCopy()
-			hostNamedRoute.Spec.Hostnames = []gatewayv1beta1.Hostname{hostname}
-			hostnamedRoutes = append(hostnamedRoutes, hostNamedRoute)
-		}
-	}
-	// split HTTPRoutes (already split once by hostname) by match.
-	newHTTPRoutes := []*gatewayv1beta1.HTTPRoute{}
-	for _, route := range hostnamedRoutes {
-		for i, rule := range route.Spec.Rules {
-			for j, match := range rule.Matches {
-				splitRoute := route.DeepCopy()
-				splitRoute.Spec.Rules = []gatewayv1beta1.HTTPRouteRule{
-					{
-						Matches:     []gatewayv1beta1.HTTPRouteMatch{match},
-						Filters:     rule.Filters,
-						BackendRefs: rule.BackendRefs,
-					},
-				}
-				if splitRoute.Annotations == nil {
-					splitRoute.Annotations = map[string]string{}
-				}
-				splitRoute.Annotations[InternalRuleIndexAnnotationKey] = strconv.Itoa(i)
-				splitRoute.Annotations[InternalMatchIndexAnnotationKey] = strconv.Itoa(j)
-				newHTTPRoutes = append(newHTTPRoutes, splitRoute)
+func SplitHTTPRoute(httproute *gatewayv1beta1.HTTPRoute) []SplitHTTPRouteMatch {
+	splitMatches := []SplitHTTPRouteMatch{}
+	splitHTTPRouteByMatch := func(hostname string) {
+		for ruleIndex, rule := range httproute.Spec.Rules {
+			for matchIndex, match := range rule.Matches {
+				splitMatches = append(splitMatches, SplitHTTPRouteMatch{
```

Comment:
Pre-allocate `splitMatches`?

EDIT: since that's a closure I'd consider 

a) moving this into for loop statement and getting rid of the closure
b) using arguments and return values in the closure

Reply:
Seems that it is not-so easy to pre-allocate it (but can also implement: iterate though all rules and matches of the source HTTPRoute to count how many matches to be split). I think it is good to use return values from the closure.

Reply:
Ah yes, it's not that easy because we'd need to rely on the fields that are nested in the slice elements 🤔

Reply:
updated the closure to return the result.